sound/soc/sof/sof-audio.c

```diff
@@ -256,29 +256,125 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
 	return 0;
 }
 
+/*
+ * free all widgets in the sink path starting from the source widget
+ * (DAI type for capture, AIF type for playback)
+ */
+static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget,
+				    int dir)
+{
+	struct snd_soc_dapm_path *p;
+	int ret;
+	int ret1 = 0;
+
+	/* free all widgets even in case of error to keep use counts balanced */
+	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
+		if (p->sink->dobj.private && widget->dobj.private) {
+			if (WIDGET_IS_AIF(widget->id) || WIDGET_IS_DAI(widget->id)) {
+				ret = sof_widget_free(sdev, widget->dobj.private);
+				if (ret < 0)
+					ret1 = ret;
+			}
+
+			ret = sof_widget_free(sdev, p->sink->dobj.private);
+			if (ret < 0)
+				ret1 = ret;
+
+			ret = sof_free_widgets_in_path(sdev, p->sink, dir);
+			if (ret < 0)
+				ret1 = ret;
+		}
+	}
+
+	return ret1;
+}
+
+/*
+ * set up all widgets in the sink path starting from the source widget
+ * (DAI type for capture, AIF type for playback).
+ * The error path in this function ensures that all successfully set up widgets getting freed.
+ */
+static int sof_set_up_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget,
+				      int dir)
+{
+	struct snd_soc_dapm_path *p;
+	int ret;
+
+	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
```

Comment:
Does this work for capture widget or should snd_soc_dapm_widget_for_each_source_path() be used (all paths out from the capture widget)?